lib/compiler/passes/generate-js.js

```diff
@@ -81,9 +81,27 @@ function wrapInSourceNode(prefix, chunk, location, suffix, name) {
 function generateJS(ast, options) {
   // These only indent non-empty lines to avoid trailing whitespace.
   function indent2(code) {
-    return code instanceof SourceNode
-      ? code
-      : code.replace(/^(.+)$/gm, "  $1");
+    let sawEol = true;
+    let inSourceNode = 0;
+    function helper(code) {
+      if (Array.isArray(code)) {
+        return code.map(helper);
+      }
+      if (code instanceof SourceNode) {
+        inSourceNode++;
+        code.children = helper(code.children);
+        inSourceNode--;
+        return code;
+      }
+      if (sawEol) {
+        code = code.replace(/^(.+)$/gm, "  $1");
+      } else {
+        code = code.replace(/\n(\s*\S)/g, "\n  $1");
+      }
+      sawEol = !inSourceNode || code.endsWith("\n");
+      return code;
+    }
```

Comment:
This code not so clear. It is better to explain it with a few lines of comments

Reply:
Will do.

Reply:
Nice!